### ISSUE TITLE
OPTIMIZER: Implement correction factor for weighted variables when 1-4 values are null

### ISSUE THREAD

**Issue by druepdx** [USFS-PNW/Fia-Biosum-Manager#368]
Null values are treated as 0 by BioSum/SQLite when summing weighted variables by rxpackage. This can cause the weighted variables to be biased low. A correction factor could be applied to adjust for the null value(s).

8 values are added to calculate the values in the prepost_fvs_weighted.db tables: cycle1pre, cycle1post, cycle2pre, cycle2post, cycle3pre, cycle3post, cycle4pre, and cycle4post. Each non null value should be multiplied by correction factor of 1/(1-sum(weights for null values).

**Comment by jsfried** [comment]
I would correct the statement to read:

8 **weighted** values are summed to generate the values in the prepost_fvs_weighted tables... That sum needs to b multiplied by a correction factor (1/(1-sum(user entered weights for Cycle/TimeInCycle cases where FVS_PREPOST table values that turn out to be null)). With the caveat that this adjustment be made only when null counts for the pre and post tables together for that stand/package combination are <= a new parameter called PrePostNullThreshold (range 1-7). When null count is > PrePostNullThreshold, then the weighted sum in the PREPOST_FVS_WEIGHTED table should be replaced with null and the analyst notified (with a dialog message or log file) so that they know to account for the possibility of missing values for that variable in the weighted tables when performing analysis in optimizer.

**Comment by druepdx** [comment]
I looked into this again, and I think it would take me about a week to a week and a half to implement.

**Comment by lbross** [comment]
Thanks! And do you see any rework you would have to do if you implemented this before the next steps of the SQLite conversion?

**Comment by druepdx** [comment]
No, it should be contained to only changing what is being written to a temporary database that's already being used.

**Comment by druepdx** [comment]
This is currently what I have for the negative values pop-up. Would anyone prefer different phrasing?
<img width="311" alt="Image" src="https://github.com/user-attachments/assets/4c41856e-5def-4167-b6d2-b4de259e7133" />

**Comment by jsfried** [comment]
That looks terrific to me. Thanks Drue!

**Comment by jsfried** [comment]
You could change the last sentence to "Click OK to close this dialog"

That is a bit odd that you'd have to do a recalc all for a new threshold to "take"-- better to place a button next to it to <Save threshold> (which pops a dialog saying the threshold saved, but not in effect until a recalc all is executed. Alternatively, the threshold could live on the project settings screen (where you set debugging levels etc.) but that is not a place folks would tend to look for it

**Comment by lbross** [comment]
Agreed on phrasing for item #1. 
If we're placing a button next to the threshold, it might as well be a save button. But we should also leave the current logic in place that saves the value before a calculation (single or recalc) is run even if the save button isn't clicked. This is the way the Processor settings work. 
Agree that no one would think to look on the project settings screen.

**Comment by jsfried** [comment]
Can the text of this dialog 1) be made at least 2 sizes larger; 2) be edited to have line breaks after: "...complete!", "determination);", "...non-null cases;" and "no missing cases"; 3) change the text to read "...**had 4 or fewer missing cases,** so..."

![Image](https://github.com/user-attachments/assets/eabc3efb-b08f-45b3-9ace-3a5cfbd5d4d0)

**Comment by lbross** [comment]
@druepdx  I know how to add the line breaks, but I'm not sure if you can change font size in a MessageBox. Can you look into this? The alternative would be to create a uc hosted in the frmDialog, but that will take more time that I'm not sure we want to spend.

**Comment by druepdx** [comment]
Yes, I will look into it. I was thinking the same thing.

**Comment by druepdx** [comment]
Like Lesley guessed, changing the font size isn't super straightforward. Here is an updated option for the layout of the text though:

<img width="313" alt="Image" src="https://github.com/user-attachments/assets/118a129a-defe-4db6-b617-635b47c40d51" />

**Comment by jsfried** [comment]
@druepdx 
Thanks, Drue-- this does look better. The font size is REALLY small-- seems smaller on this one than other dialogs but that could be my imagination.

I validated the null count calculations for Torch_Index in POTFIRE for cases (stand-package combos) where 6, and 2, Torch_Index values were null (actually, coded -1 in the PREPOST tables). Where 6 were null, the Weighted table correctly sets the value in the weighted table to null. When 2 were null, the weights were correctly adjusted in that the calculation of the weighted Torch_Index matched my hand calculations. 

Testing with a never null attribute (Summary.TPA which is 0 when no trees are cut and a positive integer otherwise) revealed that a null_count attribute is generated for EVERY weighted variable in the weighted table. Those columns are neither needed nor useful when there are no nulls. Is it problematic to include them only when one or more rows have non-zero null count for that variable? I noticed on a migratred cwi_demo project that existing weighted variables did NOT have associated null count columns-- only new variables that I created in the Define Calculated Variables workflow. When I did a recalc all, new (and superfluous given that they had no nulls) columns were created in all Weighted tables for all variables which seems to needlessly expand the size of the database without adding any value. Actually, the ABSENCE of a null count column (owing to there being no nulls in a variable) is a more usefull indicator to the analyst than a column of nulls and 0s which one has to parse or sort to assure oneself that there are no nulls (e.g., in a recalc'ed variable, given that there is no messaging to the user about null counts when recalc all is summoned). 

Let us know if it is viable to create and populate null count columns only when relevant.

Jeremy has added to his to-do list the task of documenting this capability in the BUG asap.

**Comment by druepdx** [comment]
My understanding with the message box is that it uses the default font of the system so it should be the same for all message boxes.

Regarding the null count columns, I believe I got it working to only add those columns when there are any nulls. I'm doing some more checking, but that should be viable

**Comment by jsfried** [comment]
Hmmm -- I need to learn how to vchanve the default font for the system then....

Great that you are able to thin the column count-- thanks!

**Comment by jsfried** [comment]
Dialog looks better with some white space. 

no needless columns are being created, but existing needless columns (with no positive null count values) are not removed. This can be lived with because there are not projects in the wild with needless null count columns
